File: TopCoder/src/main/java/com/leetcode/String/KeyboardRow.java
package com.leetcode.String;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class KeyboardRow {

    static Map<String, Integer> keyMap = new HashMap<>();
    static {
        keyMap.put("Q", 0);
        keyMap.put("W", 0);
        keyMap.put("E", 0);
        keyMap.put("R", 0);
        keyMap.put("T", 0);
        keyMap.put("Y", 0);
        keyMap.put("U", 0);
        keyMap.put("I", 0);
        keyMap.put("O", 0);
        keyMap.put("P", 0);

        keyMap.put("A", 1);
        keyMap.put("S", 1);
        keyMap.put("D", 1);
        keyMap.put("F", 1);
        keyMap.put("G", 1);
        keyMap.put("H", 1);
        keyMap.put("J", 1);
        keyMap.put("K", 1);
        keyMap.put("L", 1);

        keyMap.put("Z", 2);
        keyMap.put("X", 2);
        keyMap.put("C", 2);
        keyMap.put("V", 2);
        keyMap.put("B", 2);
        keyMap.put("N", 2);
        keyMap.put("M", 2);
    }


    public String[] findWords(String[] words) {

        List<String> resList = Arrays.asList(words).stream().filter(KeyboardRow::checkIfInSameRow)
                .collect(Collectors.toList());

        return resList.toArray(new String[0]);
    }

    private static boolean checkIfInSameRow(String s) {
        int currentRow = -1;
        for (int i = 0 ; i < s.length() ; i++) {
            if ( currentRow == -1) {
                currentRow = keyMap.get((s.charAt(i) +"").toUpperCase());
            } else if (currentRow != keyMap.get((s.charAt(i) +"").toUpperCase())) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        String[] words = {"Hello", "Alaska", "Dad", "Peace"};
        String[] result = new KeyboardRow().findWords(words);
        Arrays.asList(result).stream().forEach(System.out::println);
    }
}

File: TopCoder/python/misc/KeyboardRow.py


class Solution:
    def findWords(self, words):
        keyMap = {}
        keyMap["Q"] = 0
        keyMap["W"] = 0
        keyMap["E"] = 0
        keyMap["R"] = 0
        keyMap["T"] = 0
        keyMap["Y"] = 0
        keyMap["U"] = 0
        keyMap["I"] = 0
        keyMap["O"] = 0
        keyMap["P"] = 0

        keyMap["A"] = 1
        keyMap["S"] = 1
        keyMap["D"] = 1
        keyMap["F"] = 1
        keyMap["G"] = 1
        keyMap["H"] = 1
        keyMap["J"] = 1
        keyMap["K"] = 1
        keyMap["L"] = 1

        keyMap["Z"] = 2
        keyMap["X"] = 2
        keyMap["C"] = 2
        keyMap["V"] = 2
        keyMap["B"] = 2
        keyMap["N"] = 2
        keyMap["M"] = 2

        result = []

        for word in words:
            value = -1
            found = True
            for c in word:
                if value == -1:
                    value = keyMap.get(c.upper())
                elif value != keyMap.get(c.upper()):
                    found = False
                    break
            if found:
                result.append(word)

        return result

words = ["Hello", "Alaska", "Dad", "Peace"]

sol = Solution()
result = sol.findWords(words)
print(result)